File: src/vibe_kanban_clone/api/routes/tasks.py
"""Task routes."""

import asyncio
from typing import Annotated

import structlog
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession

from vibe_kanban_clone.api.deps import get_session
from vibe_kanban_clone.api.routes.ws import broadcast_project
from vibe_kanban_clone.db.engine import async_session_factory
from vibe_kanban_clone.schemas.common import PaginatedParams
from vibe_kanban_clone.schemas.task import TaskCreate, TaskMove, TaskRead, TaskRunCreate, TaskUpdate
from vibe_kanban_clone.schemas.task_comment import TaskCommentCreate, TaskCommentRead
from vibe_kanban_clone.schemas.task_run import TaskRunRead
from vibe_kanban_clone.services import comments as comments_service
from vibe_kanban_clone.services import projects as projects_service
from vibe_kanban_clone.services import runs as runs_service
from vibe_kanban_clone.services import tasks as tasks_service

router = APIRouter()
logger = structlog.get_logger()
_bg_tasks: set[asyncio.Task] = set()
_run_semaphore = asyncio.Semaphore(5)


@router.get("/projects/{project_id}/tasks", response_model=list[TaskRead])
async def list_tasks(
    project_id: int,
    pagination: Annotated[PaginatedParams, Depends()],
    session: Annotated[AsyncSession, Depends(get_session)],
) -> list[TaskRead]:
    """List all tasks in a project."""
    project = await projects_service.get_project(session, project_id)
    if project is None:
        raise HTTPException(status_code=404, detail="Project not found")
    return await tasks_service.list_tasks_by_project(
        session, project_id, pagination.limit, pagination.offset
    )


@router.post("/tasks", response_model=TaskRead, status_code=201)
async def create_task(
    data: TaskCreate,
    session: Annotated[AsyncSession, Depends(get_session)],
) -> TaskRead:
    """Create a new task."""
    task = await tasks_service.create_task(session, data)
    await broadcast_project(
        task.project_id,
        "task.created",
        TaskRead.model_validate(task).model_dump(mode="json"),
    )
    return task


@router.get("/tasks/{task_id}", response_model=TaskRead)
async def get_task(
    task_id: int,
    session: Annotated[AsyncSession, Depends(get_session)],
) -> TaskRead:
    """Get a task by ID."""
    task = await tasks_service.get_task(session, task_id)
    if task is None:
        raise HTTPException(status_code=404, detail="Task not found")
    return task


@router.patch("/tasks/{task_id}", response_model=TaskRead)
async def update_task(
    task_id: int,
    data: TaskUpdate,
    session: Annotated[AsyncSession, Depends(get_session)],
) -> TaskRead:
    """Update a task."""
    task = await tasks_service.get_task(session, task_id)
    if task is None:
        raise HTTPException(status_code=404, detail="Task not found")
    if data.expected_version is None:
        data.expected_version = task.version
    try:
        task = await tasks_service.update_task(session, task, data)
    except ValueError:
        raise HTTPException(status_code=409, detail="Task was modified by another client") from None
    await broadcast_project(
        task.project_id,
        "task.updated",
        TaskRead.model_validate(task).model_dump(mode="json"),
    )
    return task


@router.delete("/tasks/{task_id}", status_code=204)
async def delete_task(
    task_id: int,
    session: Annotated[AsyncSession, Depends(get_session)],
) -> None:
    """Delete a task."""
    task = await tasks_service.get_task(session, task_id)
    if task is None:
        raise HTTPException(status_code=404, detail="Task not found")
    project_id = task.project_id
    await tasks_service.delete_task(session, task)
    await broadcast_project(
        project_id,
        "task.deleted",
        {"task_id": task_id},
    )


@router.post("/tasks/{task_id}/move", response_model=TaskRead)
async def move_task(
    task_id: int,
    data: TaskMove,
    session: Annotated[AsyncSession, Depends(get_session)],
) -> TaskRead:
    """Move a task to another column."""
    task = await tasks_service.get_task(session, task_id)
    if task is None:
        raise HTTPException(status_code=404, detail="Task not found")
    if data.expected_version is None:
        data.expected_version = task.version
    try:
        task = await tasks_service.move_task(session, task, data)
    except ValueError:
        raise HTTPException(status_code=409, detail="Task was modified by another client") from None
    await broadcast_project(
        task.project_id,
        "task.moved",
        TaskRead.model_validate(task).model_dump(mode="json"),
    )
    return task


@router.post("/tasks/{task_id}/comments", response_model=TaskCommentRead, status_code=201)
async def create_comment(
    task_id: int,
    data: TaskCommentCreate,
    session: Annotated[AsyncSession, Depends(get_session)],
) -> TaskCommentRead:
    """Create a comment on a task."""
    task = await tasks_service.get_task(session, task_id)
    if task is None:
        raise HTTPException(status_code=404, detail="Task not found")
    comment = await comments_service.create_comment(session, task_id, data)
    await broadcast_project(
        task.project_id,
        "comment.created",
        TaskCommentRead.model_validate(comment).model_dump(mode="json"),
    )
    return comment


@router.post("/tasks/{task_id}/run", status_code=202)
async def run_task(
    task_id: int,
    data: TaskRunCreate,
    session: Annotated[AsyncSession, Depends(get_session)],
) -> dict:
    """Start a task execution run."""
    task = await tasks_service.get_task(session, task_id)
    if task is None:
        raise HTTPException(status_code=404, detail="Task not found")

    prompt = data.prompt if data.prompt is not None else (task.description or task.title)

    try:
        await asyncio.wait_for(_run_semaphore.acquire(), timeout=1.0)
    except TimeoutError:
        raise HTTPException(status_code=503, detail="Server busy, try again later") from None

    try:
        task_ref = asyncio.create_task(
            runs_service.execute_task_run(
                async_session_factory,
                task_id,
                data.model_id,
                prompt,
            )
        )
    except Exception:
        _run_semaphore.release()
        raise

    _bg_tasks.add(task_ref)

    def _on_task_done(t: asyncio.Task) -> None:
        _bg_tasks.discard(t)
        _run_semaphore.release()
        if exc := t.exception():
            logger.error("task_run_failed", error=str(exc))

    task_ref.add_done_callback(_on_task_done)

    return {"status": "started", "task_id": task_id}


@router.get("/tasks/{task_id}/runs", response_model=list[TaskRunRead])
async def list_task_runs(
    task_id: int,
    pagination: Annotated[PaginatedParams, Depends()],
    session: Annotated[AsyncSession, Depends(get_session)],
) -> list[TaskRunRead]:
    """List execution runs for a task."""
    task = await tasks_service.get_task(session, task_id)
    if task is None:
        raise HTTPException(status_code=404, detail="Task not found")
    return await runs_service.list_task_runs(session, task_id, pagination.limit, pagination.offset)

File: server/src/routes/tasks.ts
/**
 * Routes: /api/projects/:id/tasks, /api/tasks
 *   GET    /projects/:id/tasks    - list tasks for a project (paginated)
 *   POST   /tasks                 - create a new task
 *   GET    /tasks/:id             - get task detail (with comments)
 *   PATCH  /tasks/:id             - update task (409 on version mismatch)
 *   DELETE /tasks/:id             - delete task
 *   POST   /tasks/:id/move        - move task to column (409 on version mismatch)
 *   POST   /tasks/:id/comments    - add comment to task
 *   POST   /tasks/:id/run         - start LLM executor run (202 accepted)
 *   GET    /tasks/:id/runs        - list task runs (paginated)
 */
import { zValidator } from "@hono/zod-validator";
import { Hono } from "hono";
import { HTTPException } from "hono/http-exception";
import { logger } from "../lib/logger.js";
import { PaginatedParams } from "../schemas/common.js";
import {
  TaskCreate,
  TaskMove,
  TaskRunCreate,
  TaskUpdate,
} from "../schemas/task.js";
import { TaskCommentCreate } from "../schemas/taskComment.js";
import * as commentsService from "../services/comments.js";
import * as projectsService from "../services/projects.js";
import * as runsService from "../services/runs.js";
import * as tasksService from "../services/tasks.js";
import { broadcastProject } from "../ws/manager.js";

const app = new Hono();

const MAX_CONCURRENT_RUNS = 5;
let activeRuns = 0;

app.get("/projects/:id/tasks", zValidator("query", PaginatedParams), async (c) => {
  const projectId = Number(c.req.param("id"));
  const project = await projectsService.getProject(projectId);
  if (!project) throw new HTTPException(404, { message: "Project not found" });
  const { limit, offset } = c.req.valid("query");
  const tasks = await tasksService.listTasksByProject(projectId, limit, offset);
  return c.json(tasks);
});

app.post("/tasks", zValidator("json", TaskCreate), async (c) => {
  const data = c.req.valid("json");
  const task = await tasksService.createTask(data);
  await broadcastProject(task.project_id, "task.created", task);
  return c.json(task, 201);
});

app.get("/tasks/:id", async (c) => {
  const id = Number(c.req.param("id"));
  const task = await tasksService.getTask(id);
  if (!task) throw new HTTPException(404, { message: "Task not found" });
  const status = tasksService.deriveStatus(task);
  return c.json({ ...task, status });
});

app.patch("/tasks/:id", zValidator("json", TaskUpdate), async (c) => {
  const id = Number(c.req.param("id"));
  const task = await tasksService.getTask(id);
  if (!task) throw new HTTPException(404, { message: "Task not found" });
  const data = c.req.valid("json");
  if (data.expected_version === undefined) {
    data.expected_version = task.version;
  }
  try {
    const updated = await tasksService.updateTask(id, data);
    await broadcastProject(updated.project_id, "task.updated", updated);
    return c.json(updated);
  } catch (e) {
    if (e instanceof Error && e.message === "version mismatch") {
      throw new HTTPException(409, { message: "Task was modified by another client" });
    }
    throw e;
  }
});

app.delete("/tasks/:id", async (c) => {
  const id = Number(c.req.param("id"));
  const task = await tasksService.getTask(id);
  if (!task) throw new HTTPException(404, { message: "Task not found" });
  const projectId = task.project_id;
  await tasksService.deleteTask(id);
  await broadcastProject(projectId, "task.deleted", { task_id: id });
  return c.body(null, 204);
});

app.post("/tasks/:id/move", zValidator("json", TaskMove), async (c) => {
  const id = Number(c.req.param("id"));
  const task = await tasksService.getTask(id);
  if (!task) throw new HTTPException(404, { message: "Task not found" });
  const data = c.req.valid("json");
  if (data.expected_version === undefined) {
    data.expected_version = task.version;
  }
  try {
    const updated = await tasksService.moveTask(id, data);
    await broadcastProject(updated.project_id, "task.moved", updated);
    return c.json(updated);
  } catch (e) {
    if (e instanceof Error && e.message === "version mismatch") {
      throw new HTTPException(409, { message: "Task was modified by another client" });
    }
    throw e;
  }
});

app.post("/tasks/:id/comments", zValidator("json", TaskCommentCreate), async (c) => {
  const id = Number(c.req.param("id"));
  const task = await tasksService.getTask(id);
  if (!task) throw new HTTPException(404, { message: "Task not found" });
  const data = c.req.valid("json");
  const comment = await commentsService.createComment(id, data);
  await broadcastProject(task.project_id, "comment.created", comment);
  return c.json(comment, 201);
});

app.post("/tasks/:id/run", zValidator("json", TaskRunCreate), async (c) => {
  const id = Number(c.req.param("id"));
  const task = await tasksService.getTask(id);
  if (!task) throw new HTTPException(404, { message: "Task not found" });
  const data = c.req.valid("json");

  if (activeRuns >= MAX_CONCURRENT_RUNS) {
    throw new HTTPException(503, { message: "Server busy, try again later" });
  }

  const prompt = data.prompt ?? (task.description || task.title);
  activeRuns++;

  runsService
    .executeTaskRun(id, data.model_id, prompt)
    .catch((err) => logger.error({ err }, "task_run_failed"))
    .finally(() => {
      activeRuns--;
    });

  return c.json({ status: "started", task_id: id }, 202);
});

app.get("/tasks/:id/runs", zValidator("query", PaginatedParams), async (c) => {
  const id = Number(c.req.param("id"));
  const task = await tasksService.getTask(id);
  if (!task) throw new HTTPException(404, { message: "Task not found" });
  const { limit, offset } = c.req.valid("query");
  const runs = await runsService.listTaskRuns(id, limit, offset);
  return c.json(runs);
});

export default app;
